lib/matplotlib/tests/test_backend_tk.py

```diff
@@ -1,236 +1,181 @@
+import functools
+import inspect
 import os
+import re
 import subprocess
 import sys
 
 import pytest
 
 _test_timeout = 10  # Empirically, 1s is not enough on CI.
 
-# NOTE: TkAgg tests seem to have interactions between tests,
-# So isolate each test in a subprocess. See GH#18261
+
+def _isolated_tk_test(success_count, func=None):
+    """
+    A decorator to run *func* in a subprocess and assert that it prints
+    "success" *success_count* times and nothing on stderr.
+
+    TkAgg tests seem to have interactions between tests, so isolate each test
+    in a subprocess. See GH#18261
+
+    The decorated function must be fully self-contained, and thus perform
+    all the imports it needs.  Because its source is extracted and run by
+    itself, coverage will consider it as not being run, so it should be marked
+    with ``# pragma: no cover``
+    """
+
+    if func is None:
+        return functools.partial(_isolated_tk_test, success_count)
+
+    # Remove decorators.
+    source = re.search(r"(?ms)^def .*", inspect.getsource(func)).group(0)
+
+    def test_func():
+        try:
+            proc = subprocess.run(
+                [sys.executable, "-c", f"{source}\n{func.__name__}()"],
+                env={**os.environ, "MPLBACKEND": "TkAgg"},
+                timeout=_test_timeout,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                check=True,
+                universal_newlines=True,
+            )
+        except subprocess.TimeoutExpired:
+            pytest.fail("Subprocess timed out")
+        except subprocess.CalledProcessError:
+            pytest.fail("Subprocess failed to test intended behavior")
+        else:
+            assert proc.stdout.count("success") == success_count
+            # macOS may actually emit irrelevant errors about Accelerated
+            # OpenGL vs. software OpenGL, so suppress them.
+            assert not [line for line in proc.stderr.splitlines()
+                        if "OpenGL" not in line]
+
+    test_func.__name__ = func.__name__
```

Comment:
Why not use `functools.wraps`?

Reply:
I wasn't really sure if that semantically made sense as "wrapping" but I guess it does.